### PR TITLE
An implemenation of the ITU-R BS.1770-4 loudness recommendation

### LINKED ISSUE #1205: Add ITU-R BS.1770-4 Loudness
## 🚀 Feature

Implement a standard algorithm for computing audio loudness, like ITU-R BS.1770-4.

## Motivation

Audio loudness is an important feature for representing audio. For example, it's used in this popular library: https://github.com/magenta/ddsp

## Pitch

Here is a `numpy` implementation of ITU-R BS.1770-4. It'd be helpful to have this translated to PyTorch and included in this library: https://github.com/csteinmetz1/pyloudnorm/blob/master/pyloudnorm/meter.py

Here is the official loudness implementation from the original paper: https://github.com/BrechtDeMan/loudness.py/blob/master/loudness.py

### PR DESCRIPTION
**Pull request by bshall**
I took a stab at implementing the ITU-R BS.1770-4 loudness recommendation (closes #1205). To give some more details:
- I've implemented K-weighting following csteinmetz1 instead of BrechtDeMan since it fit well with torchaudio's already implemented filters (`treble_biquad` and `highpass_biquad`).
- I've added four audio files to test compliance with the recommendation. These are linked in [this pdf](https://www.itu.int/dms_pub/itu-r/opb/rep/R-REP-BS.2217-2-2016-PDF-E.pdf). There are many more test files there but I didn't want to bog down the assets directory with too many files. Let me know if I should add or remove anything.
- I've kept many of the constant internal to the function (e.g. the block duration, overlap, and the absolute threshold gamma). I'm not sure if these should be exposed in the signature.
- I've implemented support for up to 5 channels (following both csteinmetz1 and BrechtDeMan). The recommendation includes weights for up to 24 channels. Is there any convention for how many channels to support?

I hope this is helpful! looking forward to hearing from you.

### CODE REVIEW

**Review by mthrok — COMMENTED** [review]
Hi @bshall 

Thanks for the PR. 
Quick questions before looking into the PR:
1. Where are the reference files from? I can get the article https://www.itu.int/pub/R-REP-BS.2217-2-2016, but cannot find links to ~the data~ (EDIT: PDF contains the link to the data) nor the license information. If the license is unclear, we cannot checkin the files. In that case, we need to work around by fetching data from a web at test runtime.
2. Currently torchaudio main branch is 14.83 MB. The assets will double the size of main branch. Is there a way to reduce the asset size?

**Comment by bshall** [comment]
Hi @mthrok, I did some digging around and couldn't find a license on the ITU website so I rewrote the tests to download the files at runtime. I created a new test class but can add the tests to the `Functional` class if that's preferable. I've also squashed the commits so there isn't any history of the files in the branch. Also, this way we don't have to worry about increasing the size of assets :)

**Comment by bshall** [comment]
Hi @mthrok, just wanted to follow up on this pull request since there hasn't been an update in a little while. I've rebased on main so it should be up to date.

**Comment by mthrok** [comment]
Hi @bshall 

Sorry I had been busy regarding the recent release. I will have time to review this soon.
One thing about the testing, one of the restrictions on unit test is that the test should not rely on external resource. Since now that the assets are fetched on-the-fly, can we move the test to integration directory?

**Comment by mthrok** [comment]
@carolineechen Can you help review?

**Review by carolineechen — COMMENTED** [review]
Thanks for working on this! Left a couple of comments, mostly small code suggestions and the two main things are
- moving the test to integration_tests bc of reliance on external source
- adding this function to transforms as well

File: torchaudio/functional/functional.py [hunk 1]
```diff
@@ -1602,6 +1604,66 @@ def edit_distance(seq1: Sequence, seq2: Sequence) -> int:
...
+    .. devices:: CPU CUDA
+
+    .. properties:: TorchScript
+
+    Args:
+        waveform(torch.Tensor): audio waveform of dimension of `(..., channels, time)`
+        sample_rate (int): sampling rate of the waveform, e.g. 44100 (Hz)
+
+    Returns:
+        Tensor: loudness estimates (LKFS)
```
[the remark this answers is not in the document]
Comment (bshall) [reply torchaudio/functional/functional.py:1619]: Currently the function works on batched inputs as well so I'm returning a Tensor of LKFS values to deal with this case.

File: torchaudio/functional/functional.py [hunk 2]
```diff
@@ -1602,6 +1604,66 @@ def edit_distance(seq1: Sequence, seq2: Sequence) -> int:
...
+    Reference:
+        - https://www.itu.int/rec/R-REC-BS.1770-4-201510-I/en
+    """
+
+    if waveform.size(-2) > 5:
+        raise ValueError("Only up to 5 channels are supported.")
+
+    gate_duration: float = 0.4
+    overlap: float = 0.75
+    gamma_abs: float = -70.0
```
[the remark this answers is not in the document]
Comment (bshall) [reply torchaudio/functional/functional.py:1630]: I'm not sure if there's a potential use case for adjusting any of the parameters. I've always just stuck to the settings in the recommendation and I'm sure that covers most of the usage. On the other hand it's pretty easy to expose these values so maybe it's worth doing just for the flexibility?

File: test/torchaudio_unittest/functional/loudness_compliance_test_impl.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    "1770-2_Comp_AbsGateTest": "http://www.itu.int/dms_pub/itu-r/oth/11/02/R11020000010029ZIPM.zip",
+    "1770-2_Comp_24LKFS_500Hz_2ch": "http://www.itu.int/dms_pub/itu-r/oth/11/02/R11020000010018ZIPM.zip",
+    "1770-2 Conf Mono Voice+Music-24LKFS": "http://www.itu.int/dms_pub/itu-r/oth/11/02/R11020000010038ZIPM.zip",
+}
+
+
+class Loudness(TempDirMixin, TestBaseMixin):
+    def download_and_extract_file(self, filename):
+        zippath = self.get_temp_path(filename + ".zip")
+        torch.hub.download_url_to_file(_COMPLIANCE_FILE_URLS[filename], zippath, progress=False)
```
[the remark this answers is not in the document]
Comment (bshall) [reply test/torchaudio_unittest/functional/loudness_compliance_test_impl.py]: I'm not sure how I'd use download_asset from torchaudio.utils. It looks like the key has to be relative to https://download.pytorch.org/torchaudio/. Let me know if I'm missing something.

File: torchaudio/transforms/_transforms.py [hunk 4]
```diff
@@ -1248,6 +1248,32 @@ def __init__(self, time_mask_param: int, iid_masks: bool = False, p: float = 1.0
...
+
+    .. devices:: CPU CUDA
+
+    .. properties:: TorchScript
+
+    Reference:
+        - https://www.itu.int/rec/R-REC-BS.1770-4-201510-I/en
+    """
+
+    def __init__(self):
```
Comment (carolineechen) [inline torchaudio/transforms/_transforms.py:1262]: let's move the `sample_rate` parameter into `__init__`, set `self.sample_rate`, and move the docstrings up too

File: test/torchaudio_unittest/functional/loudness_compliance_test_impl.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (carolineechen) [reply test/torchaudio_unittest/functional/loudness_compliance_test_impl.py]: this is my bad, we can keep the original `torch.hub.download_url_to_file`, thanks for catching this

File: torchaudio/functional/functional.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (carolineechen) [reply torchaudio/functional/functional.py]: got it, in that case let's keep it defined in the function for now. and if we get a feature request with a use case in the future asking to expose any of the parameters, we can easily address it then.

File: test/integration_tests/loudness_compliance_test_impl.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+class Loudness(TempDirMixin, TestBaseMixin):
+    def download_and_extract_file(self, filename):
+        zippath = self.get_temp_path(filename + ".zip")
+        torch.hub.download_url_to_file(_COMPLIANCE_FILE_URLS[filename], zippath, progress=False)
+        with zipfile.ZipFile(zippath) as file:
+            file.extractall(os.path.dirname(zippath))
+        return self.get_temp_path(filename + ".wav")
+
+    def test_measure_loudness_relative_gate(self):
```
Comment (carolineechen) [inline test/integration_tests/loudness_compliance_test_impl.py:26]: looks like these tests just differ in the file path name and the expected tensor. can you reduce it to one test function and test for the different samples using `parameterized`?

```
parameterized.expand(
    [
        ("1770-2_Comp_RelGateTest", -10.0),
        ("1770-2_Comp_AbsGateTest", -69.5),
         ...
    ]
)
def test_measure_loudness(self, filepath, expected):
    ....
```

**Review by carolineechen — COMMENTED** [review]
thanks for the fixes and responses, a few more things

**Comment by carolineechen** [comment]
Could you also rebase onto main (`build_docs` job on CircleCI should be resolved), and also add the appropriate docs entry into `functional.rst` and `transforms.rst`?

**Comment by bshall** [comment]
Hi @carolineechen, thanks for the feedback! I've made all the requested changes and rebased on main so hopefully everything is in good shape now. Let me know if there's anything else.

**Comment by carolineechen** [comment]
Changes look good, thanks for working on this! Integration test on CircleCI is failing though, think you may need to add `expecttest` and `parameterized` to the integration test workflow [here](https://github.com/pytorch/audio/blob/main/.github/workflows/integration-test.yml#L32)

**Comment by bshall** [comment]
@carolineechen, I've added `scipy` to the integration test workflow as well. Hopefully that's the last requirement.

**Comment by carolineechen** [comment]
@bshall so sorry for all the complications with the Github build/test, there was an upstream change from PyTorch so unittests on older PRs are now breaking, any chance you could rebase again, or I'm happy to take over resolving the CircleCI testing if that is easier for you. thanks again for the contributions!

**Comment by mthrok** [comment]
> Changes look good, thanks for working on this! Integration test on CircleCI is failing though, think you may need to add `expecttest` and `parameterized` to the integration test workflow [here](https://github.com/pytorch/audio/blob/main/.github/workflows/integration-test.yml#L32)

@carolineechen We should not be using unit test fixtures in integration test. Integration tests should run without such harness. This PR is okay, but we need to get rid of unit test fixtures from integration test.

**Comment by bshall** [comment]
Hi @carolineechen, no problem! I'm happy to see this one through, but if its taking too long for me to make the changes and you want to jump in that's cool. I've rebased so hopefully the tests will pass now and there won't be any other issues to resolve :)

**Comment by mthrok** [comment]
@bshall `torchaudio_unittest` module cannot be used in integration test. Can you either remove it or, mark this PR editable by maintainers so that we can push on top of your commits?

**Comment by bshall** [comment]
@mthrok and @carolineechen, I've removed the `torchaudio_unittest` module from the tests and also checked "allow edits by maintainers" for the branch. I guess since this is just an integration test we don't need to worry about testing different `dtypes` or `devices`? Not sure how to do that without importing `torchaudio_unittest`.

**Comment by carolineechen** [comment]
[stylecheck](https://app.circleci.com/pipelines/github/pytorch/audio/12014/workflows/f31c6f2f-a0dc-45f4-8414-7a624f9d7803/jobs/829527) is failing, could you make that small fix, and then I think the rest looks good. Thanks for the changes!

**Review by carolineechen — APPROVED** [review]

**Pull request closed without merging**